chain/store/splitstore/splitstore.go

```diff
@@ -0,0 +1,414 @@
+}
+
+func (s *SplitStore) Has(cid cid.Cid) (bool, error) {
+	has, err := s.hot.Has(cid)
+
+	if err != nil {
+		return false, err
+	}
+
+	if has {
+		return true, nil
+	}
+
+	return s.cold.Has(cid)
+}
+
+func (s *SplitStore) Get(cid cid.Cid) (blocks.Block, error) {
+	blk, err := s.hot.Get(cid)
+
+	switch err {
+	case nil:
+		return blk, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.Get(cid)
+
+	default:
+		return nil, err
+	}
+}
+
+func (s *SplitStore) GetSize(cid cid.Cid) (int, error) {
+	size, err := s.hot.GetSize(cid)
+
+	switch err {
+	case nil:
+		return size, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.GetSize(cid)
+
+	default:
+		return 0, err
+	}
+}
+
+func (s *SplitStore) Put(blk blocks.Block) error {
+	epoch := s.curTs.Height()
+	err := s.snoop.Put(blk.Cid(), epoch)
+	if err != nil {
+		return err
+	}
+
+	return s.hot.Put(blk)
+}
+
+func (s *SplitStore) PutMany(blks []blocks.Block) error {
+	err := s.hot.PutMany(blks)
+	if err != nil {
+		return err
+	}
+
+	epoch := s.curTs.Height()
+
+	batch := make([]cid.Cid, 0, len(blks))
+	for _, blk := range blks {
+		batch = append(batch, blk.Cid())
+	}
+
+	return s.snoop.PutBatch(batch, epoch)
+}
+
+func (s *SplitStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	chHot, err := s.hot.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	chCold, err := s.cold.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer cancel()
+		defer close(ch)
+
+		for _, in := range []<-chan cid.Cid{chHot, chCold} {
+			for cid := range in {
+				select {
+				case ch <- cid:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+func (s *SplitStore) HashOnRead(enabled bool) {
+	s.hot.HashOnRead(enabled)
+	s.cold.HashOnRead(enabled)
+}
+
+func (s *SplitStore) View(cid cid.Cid, cb func([]byte) error) error {
+	err := s.hot.View(cid, cb)
+	switch err {
+	case bstore.ErrNotFound:
+		return s.cold.View(cid, cb)
+
+	default:
+		return err
+	}
+}
+
+// State tracking
+func (s *SplitStore) Start(cs *store.ChainStore) error {
+	s.cs = cs
+	s.curTs = cs.GetHeaviestTipSet()
+
+	// load base epoch from metadata ds
+	// if none, then use current epoch because it's a fresh start
+	bs, err := s.ds.Get(baseEpochKey)
+	switch err {
+	case nil:
+		s.baseEpoch = bytesToEpoch(bs)
+
+	case dstore.ErrNotFound:
+		err = s.setBaseEpoch(s.curTs.Height())
+		if err != nil {
+			return err
+		}
+
+	default:
+		return err
+	}
+
+	// watch the chain
+	cs.SubscribeHeadChanges(s.HeadChange)
+
+	return nil
+}
+
+func (s *SplitStore) Close() error {
+	if s.isCompacting() {
+		log.Warn("ongoing compaction; waiting for it to finish...")
+		for s.isCompacting() {
+			time.Sleep(time.Second)
+		}
+	}
+
+	return s.env.Close()
+}
+
+func (s *SplitStore) HeadChange(revert, apply []*types.TipSet) error {
+	s.curTs = apply[len(apply)-1]
```

Comment:
This is racy for two reasons.
1. Easy one: the data race. Suggestion: turn this into an atomic.Value.
2. Harder one: I _think_ the head change notification is asynchronous, so we might start receiving Puts and Gets before we update `s.curTs`.

@Kubuxu ideas to solve (2)?

Reply:
There is no data race -- reorg notifications from the chainstore are _synchornous_.

Reply:
if you refer to the (1) data race as the assignment and the (single) read in put, then it's entirely inconsequential.
We also have the state mutex we can use if you fear a race detector detecing nonsense, but still.

Reply:
That makes unsafe assumptions about the rest of the system. And those assumptions can change any time, and it will not be noticed because there is not contract/interface breakage. Thus it will lead to a bug. Highly encourage you to synchronise here, by whichever means you want. An atomic is faster.

Reply:
It's a pointer read/write -- this things don't go in half, but fine.

Reply:
done, I used the mutex.